sdks/python/apache_beam/io/gcp/pubsub.py

```diff
@@ -290,21 +290,26 @@ def __init__(self,
         topic, id_label, with_attributes, timestamp_attribute)
 
   @staticmethod
-  def to_proto_str(element):
+  def message_to_proto_str(element):
     # type: (PubsubMessage) -> bytes
     if not isinstance(element, PubsubMessage):
       raise TypeError(
           'Unexpected element. Type: %s (expected: PubsubMessage), '
           'value: %r' % (type(element), element))
     return element._to_proto_str()
 
+  @staticmethod
+  def bytes_to_proto_str(element):
+    # type: (bytes) -> bytes
+    msg = pubsub.types.pubsub_pb2.PubsubMessage()
+    msg.data = element
+    return msg.SerializeToString()
+
   def expand(self, pcoll):
     if self.with_attributes:
```

Comment:
This is fine but probably means that we'll have to keep the with_attributes property in PubSubWritePayload to properly reconstruct the transforms from the proto.

Reply:
We can also clean up `with_attributes` and check the element type in `Map`. It depends on which way is preferred. One pushback for checking element type might be it will allow the upstream to produce a mixed pcollection of `PubsubMessage` and `bytes`.

Reply:
Using the Write transform that consumes the Pubsub proto after formatting should be the well known transform represented by the PubSubWritePayload and not the WriteToPubSub transform as it contains an ambiguous input type.

Reply:
After a video meeting, we should update the Dataflow translation for v2 to always pass through the `has_attributes` version for the sink part (aka Write transform in Python)

Reply:
Lemme look into this.

Reply:
Sent https://github.com/apache/beam/pull/12888